### PR TITLE
[autograd] Fix NaN in higher-order gradients of logaddexp / logaddexp2

### LINKED ISSUE #196704: torch.logaddexp produces NaN in second-order AD for finite logits
### 🐛 Describe the bug

`torch.logaddexp` returns `nan` for a nested reverse-mode second derivative, even though every logit is finite and the mathematical second derivative is finite.
Each `logaddexp` term is smooth; its curvature can be written with finite logistic weights. The claim is the nonfinite autodiff result, not agreement with a tiny last-bit value. At `t = -800.0` the forward value is `10003.0` and matches the reference. The second backward pass returns `nan`.
This issue reproduces on CPU, so CUDA availability does not affect the result.
```python
import torch
dtype = torch.float64
t = torch.tensor(-800.0, dtype=dtype, requires_grad=True)
x = torch.stack((t + 2, -2 * t)).reshape(2, 1)
y = torch.stack((t / 2 - 1, 3 - t, 2 * t + 1)).reshape(1, 3)
w = torch.tensor([[1, 2, -1], [3, -2, 4]], dtype=dtype)
out = (torch.logaddexp(x, y) * w).sum()
g, = torch.autograd.grad(out, t, create_graph=True)
actual, = torch.autograd.grad(g, t)
print("actual:")
print(actual)
print("expected:")
print("finite")
```
### Actual vs expected result
```text
actual:
nan
expected:
finite
```
The discrepancy occurs in the nested (second-order) backward pass. The forward value is `10003.0`. The second derivative should stay finite for these finite logits; PyTorch returns `nan`.

### Versions

```text
PyTorch version: 2.12.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A
OS: Ubuntu 24.04.3 LTS (x86_64)
Python version: 3.13.13 | packaged by Anaconda, Inc.
Python platform: Linux x86_64
Is CUDA available: False
GPU models:
  GPU 0: NVIDIA RTX 6000 Ada Generation
  GPU 1: NVIDIA RTX 6000 Ada Generation
Nvidia driver version: 570.211.01
CPU:
  Model name: AMD Ryzen Threadripper PRO 7985WX 64-Cores
  CPU(s): 128
Versions of relevant libraries:
[pip3] numpy==2.4.6
[pip3] torch==2.12.0
[pip3] triton==3.7.0
```

### PR DESCRIPTION
**Pull request by LiRunGuo**
## Issue

Fixes #196704

## Summary

`logaddexp` / `logaddexp2` backward is the logistic weight, but it was written
as `1 / (1 + exp(other - self))` (and `1 / (1 + pow(2, other - self))`). Those
overflow to `inf` once `other - self` exceeds `log(DBL_MAX) ~= 709` (`1024` for
base 2). The first-order gradient is unaffected (`grad / inf == 0`, the correct
rounding there), but the expression autograd differentiates for a double
backward evaluates to `inf / inf == nan`, so second-order reverse-mode AD
returned `nan` for finite inputs. The same expression drives the forward-mode
(`result:`) formula, so nested `torch.func.jvp` was broken the same way. The
formulas now use `sigmoid`, which is bounded and whose backward is defined in
terms of its output, so every higher-order derivative stays finite.

## AI assistance disclosure

> **AI-generated content (contained).** AI coding assistants were used for:
> (1) the root-cause analysis that `exp(other - self)` overflows and that
> differentiating it yields `inf / inf` in the double backward, (2) the proposed
> rewrite in terms of `sigmoid` shown in the diff, (3) the tests in
> `test/test_binary_ufuncs.py`, and (4) the evidence table in the collapsed
> section below. Summary of the change: "express the logistic weight with the
> bounded primitive `sigmoid` so the autograd decomposition of the backward
> cannot overflow; the first-order weight is algebraically unchanged (ATen's
> sigmoid is `1 / (1 + exp(-x))`, the same denominator), so only last-ulp
> rounding moves, measured below."

**Human commentary:** I reproduced the issue locally and confirmed the
higher-order gradients become finite; I accept the saturation beyond
`|gap| ~= 37` since it matches sigmoid's own derivative.

## Checklist

- [x] Passes lint (`spin fixlint`; verified locally with `lintrunner -a`: `ok No lint issues`)
- [x] Added/updated tests
- [ ] Updated documentation (not applicable; no public API or documented behaviour change)
- [ ] Included benchmark results (not applicable; no perf-sensitive change)

## BC-breaking?

No. The first-order weight is algebraically unchanged; only floating-point
rounding moves (`grad * sigmoid(x)` instead of `grad / (1 + exp(-x))`, and
`2**x` is now evaluated as `exp(x * ln 2)`), and second-order results become
finite instead of `nan`. Measured against the old formulas on the fixed build
(400k random inputs per dtype, operand gaps within ±1600, CPU and CUDA):
f32/f64 first-order values differ by at most a few ulps in normal ranges, with
larger relative deviations only on outputs that are denormal or underflow
(`|grad| <~ 1e-38` for f32, `<~ 1e-300` for f64); fp16/bf16 differ by up to one
ulp at large `|grad|`, and where the old half-precision `1 + exp(...)`
overflowed to `inf` the fp16 result is now the correct tiny value instead of
`0`; complex agrees to rounding.
Known accuracy characteristic for reviewers: evaluating `sigma'(x)` as
`sigma(x) * (1 - sigma(x))` saturates to `0` for `|x| >~ 37`, which is exactly
how `torch.sigmoid`'s own derivatives behave (measured), so the band
`37 < |gap| < 709` reports `0` instead of a value `<= 8.5e-17`, and everything
past the old overflow threshold is finite instead of `nan`.

<details>
<summary>Test plan and evidence</summary>

Source build `2.15.0a0+gita4a01d1`, run on CPU and on an H200 (CUDA 12.8,
`TORCH_CUDA_ARCH_LIST=9.0`).

| command | before | after |
| --- | --- | --- |
| `python test/test_binary_ufuncs.py -k logaddexp_higher_order` (new test) | `FAILED: logaddexp: second derivative is not finite for gap=800.0` | 4 passed (f32/f64, CPU + CUDA) |
| `python test/test_binary_ufuncs.py -k logaddexp` | 108 passed | 239 passed, 5 skipped |
| `python test/test_ops.py -k logaddexp` (gradcheck, gradgradcheck, forward AD, fwgrad_bwgrad, complex) | not run | 248 passed, 31 skipped, 13 expected failures (all pre-existing `_refs` complex markers) |
| `lintrunner -a tools/autograd/derivatives.yaml test/test_binary_ufuncs.py` | - | `ok No lint issues` |

Reproducer from the issue (second derivative at `t = -800`): `nan` before,
`1.92336048244828e-173` after, identical on CPU and CUDA.

First-order rounding re-check (400k random inputs per dtype, gaps within
±1600, CPU + CUDA): f32/f64 vs the old formulas differ ≤ a few ulps in normal
ranges; the larger relative deviations (up to ~1e-3 f32 / ~1e-13 f64) appear
only where the result itself is denormal or underflows to zero; f16/bf16
differences are within one ulp at large gradients (and no longer flush to `0`
where the old half-precision denominator overflowed).

</details>

### CODE REVIEW

**Comment by LiRunGuo** [comment]
Friendly ping: no reviewer has been assigned here and there's been no response for more than 4 business days. CI is green (99 passing, 1 neutral).

@albanD @soulitzer, could one of you take a look or route this to the right reviewer? It's two files: `tools/autograd/derivatives.yaml` rewrites the logaddexp / logaddexp2 backward and forward formulas in terms of `sigmoid()` instead of `1 / (1 + exp(other - self))`, which overflowed to `inf` for large operand gaps and made second-order AD return `nan`; `test/test_binary_ufuncs.py` adds a finiteness regression plus an accuracy check against `sigma'(a - b)`. Fixes #196704.
